### ISSUE TITLE
Wrap all extern C functions in catch_unwind

### ISSUE THREAD

**Issue by rukai** [gfx-rs/wgpu#233]
Another issue arising from discussion in the rust community discord.

Panicking over FFI boundaries is UB, so we should wrap every extern C function with catch_unwind.

Ideally the rust API could avoid the cost of this, but any solution there would still involve catch unwind'ing in the extern C functions.

**Comment by Lokathor** [comment]
yes. to continue a bit more on how pure rust can avoid it: split any extern function into a rust impl and also a `#[no_mangle]` and `extern "C"` wrapper that just calls over to the rust impl. Rust code can just call the impl directly and avoid getting the C abi in its call stack.

this is apparently done in some places already, it just needs to be put in all places and watched for carefully as the lib grows

**Comment by paulkernfeld** [comment]
If we do catch a panic, what's the right thing to do? Print a stack trace and abort? I see that #638 exists, which should cover most errors, but I imagine that it could still make sense to `catch_unwind` panics caused by programming errors in the `wgpu` stack.

**Comment by kvark** [comment]
Note that panics inside `wgpu` are not in our longer term goals. See https://github.com/gfx-rs/wgpu/issues/638
Sp doing this work now will likely be a wasteful effort.
